src/server/pegasus_hotkey_collector.h

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2017, Xiaomi, Inc.  All rights reserved.
+// This source code is licensed under the Apache License Version 2.0, which
+// can be found in the LICENSE file in the root directory of this source tree.
+
+#pragma once
+
+#include <dsn/cpp/serverlet.h>
+#include <dsn/utility/error_code.h>
+#include <rrdb/rrdb_types.h>
+#include <gtest/gtest_prod.h>
+#include <dsn/utility/ringbuf.h>
+
+namespace pegasus {
+namespace server {
+
+class hotkey_collector
+{
+public:
+    hotkey_collector() : _collector_state(STOP), _coarse_result(-1)
+    {
+        kMaxTime_sec = dsn_config_get_value_uint64(
+            "pegasus.server",
+            "_hotkey_collector_max_exist_time",
+            45,
+            "_hotkey_collector_max_exist_time, after that the collection will stop automatically");
+    }
+
+    bool init()
+    {
+        if (_collector_state.load(std::memory_order_seq_cst) == COARSE ||
+            _collector_state.load(std::memory_order_seq_cst) == FINE) {
+            return false;
+        }
+        if (_collector_state.load(std::memory_order_seq_cst) == FINISH) {
+            clear();
+        }
+        _timestamp = dsn_now_s();
+        _collector_state.store(COARSE, std::memory_order_seq_cst);
+        return true;
+    }
+
+    void clear()
+    {
+        for (int i = 0; i < 103; i++) {
+            _coarse_count[i].store(0, std::memory_order_seq_cst);
+            _fine_capture_unit[i].mutex.lock();
+            while (!_fine_capture_unit[i].queue.empty()) {
+                _fine_capture_unit[i].queue.pop();
+            }
+            _fine_capture_unit[i].mutex.unlock();
+        }
+        _coarse_result.store(-1, std::memory_order_seq_cst);
+        _fine_result = "";
+        _fine_count.clear();
+        _collector_state.store(STOP, std::memory_order_seq_cst);
+    }
+
+    void capture_blob_data(const ::dsn::blob &key);
+    void capture_msg_data(dsn::message_ex **requests_point, const int count);
+    void capture_str_data(const std::string &data);
+
+    void analyse_data();
+
+    std::string get_status()
+    {
+        collector_state_set status = _collector_state.load(std::memory_order_seq_cst);
+        if (status == STOP)
+            return "STOP";
+        if (status == COARSE)
+            return "COARSE";
+        if (status == FINE)
+            return "FINE";
+        return "FINISH";
+    }
+
+    bool get_result(std::string &result)
+    {
+        if (_collector_state.load(std::memory_order_seq_cst) != FINISH)
+            return false;
+        result = _fine_result;
+        return true;
+    }
+
+    enum collector_state_set
+    {
+        STOP = 0,
+        COARSE,
+        FINE,
+        FINISH
+    };
+
+private:
+    void capture_coarse_data(const std::string &data);
+    void capture_fine_data(const std::string &data);
+    const int analyse_coarse_data();
+    bool analyse_fine_data();
+
+    std::atomic<collector_state_set> _collector_state;
```

Comment:
coarse analysis and fine analysis has independent structures and functions, better to separet into two embeded classes.